File: backend/pkg/db/pocketbase.go
package db

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// PocketBaseClient es un cliente HTTP para comunicarse con PocketBase.
type PocketBaseClient struct {
	BaseURL    string
	HTTPClient *http.Client
	AdminToken string
}

// UserRecord representa un registro de usuario de PocketBase.
type UserRecord struct {
	ID   string `json:"id"`
	Role string `json:"role"`
}

// NewPocketBaseClient crea una nueva instancia del cliente PocketBase.
func NewPocketBaseClient(baseURL string) *PocketBaseClient {
	return &PocketBaseClient{
		BaseURL: baseURL,
		HTTPClient: &http.Client{
			Timeout: 10 * time.Second,
		},
	}
}

// AuthAdmin autentica como administrador y almacena el token.
func (c *PocketBaseClient) AuthAdmin(email, password string) error {
	endpoint := fmt.Sprintf("%s/api/admins/auth-with-password", c.BaseURL)

	payload := map[string]string{
		"identity": email,
		"password": password,
	}
	jsonData, _ := json.Marshal(payload)

	req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("fallo la autenticación de admin, status: %d", resp.StatusCode)
	}

	var result struct {
		Token string `json:"token"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return err
	}

	c.AdminToken = result.Token
	return nil
}

// AuthUser autentica un usuario final contra la colección "users" de PocketBase.
// Retorna el token JWT, los datos del usuario, y un error si falla.
func (c *PocketBaseClient) AuthUser(email, password string) (string, *UserRecord, error) {
	endpoint := fmt.Sprintf("%s/api/collections/users/auth-with-password", c.BaseURL)

	payload := map[string]string{
		"identity": email,
		"password": password,
	}
	jsonData, _ := json.Marshal(payload)

	req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		return "", nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return "", nil, fmt.Errorf("error conectando con PocketBase: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", nil, fmt.Errorf("credenciales inválidas, status: %d", resp.StatusCode)
	}

	var result struct {
		Token  string     `json:"token"`
		Record UserRecord `json:"record"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", nil, fmt.Errorf("error decodificando respuesta: %w", err)
	}

	return result.Token, &result.Record, nil
}

// ValidateToken valida un token JWT de usuario refrescándolo contra PocketBase.
// Retorna los datos del usuario si el token es válido.
func (c *PocketBaseClient) ValidateToken(token string) (*UserRecord, error) {
	endpoint := fmt.Sprintf("%s/api/collections/users/auth-refresh", c.BaseURL)

	req, err := http.NewRequest("POST", endpoint, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", token)

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("error conectando con PocketBase: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("token inválido, PocketBase respondió status %d", resp.StatusCode)
	}

	var result struct {
		Record UserRecord `json:"record"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("error decodificando respuesta: %w", err)
	}

	return &result.Record, nil
}

// CreateUser crea un nuevo usuario en la colección "users".
func (c *PocketBaseClient) CreateUser(email, password, passwordConfirm, name, role string) (*UserRecord, error) {
	endpoint := fmt.Sprintf("%s/api/collections/users/records", c.BaseURL)

	payload := map[string]string{
		"email":           email,
		"password":        password,
		"passwordConfirm": passwordConfirm,
		"name":            name,
		"role":            role,
	}
	jsonData, _ := json.Marshal(payload)

	req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	// Generalmente crear usuario no requiere admin token si la colección está abierta, 
	// pero si se requiere, deberíamos pasarlo. Asumiendo que pocketbase maneja la validación.

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("error conectando con PocketBase: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNoContent {
		bodyBytes, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("error creando usuario, status: %d, response: %s", resp.StatusCode, string(bodyBytes))
	}

	var record UserRecord
	if err := json.NewDecoder(resp.Body).Decode(&record); err != nil {
		return nil, fmt.Errorf("error decodificando respuesta: %w", err)
	}

	return &record, nil
}

// GetRecord obtiene un registro específico de una colección.
func (c *PocketBaseClient) GetRecord(collection, id string) ([]byte, error) {
	endpoint := fmt.Sprintf("%s/api/collections/%s/records/%s", c.BaseURL, collection, id)

	req, err := http.NewRequest("GET", endpoint, nil)
	if err != nil {
		return nil, err
	}

	if c.AdminToken != "" {
		req.Header.Set("Authorization", c.AdminToken)
	}

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	return io.ReadAll(resp.Body)
}

// ListRecords obtiene una lista paginada de registros de una colección.
func (c *PocketBaseClient) ListRecords(collection string, page, limit int, filter string) ([]byte, error) {
	endpoint := fmt.Sprintf("%s/api/collections/%s/records?page=%d&perPage=%d", c.BaseURL, collection, page, limit)
	if filter != "" {
		endpoint += "&filter=" + filter // Simplificado, debería usar url.QueryEscape en un caso real
	}

	req, err := http.NewRequest("GET", endpoint, nil)
	if err != nil {
		return nil, err
	}

	if c.AdminToken != "" {
		req.Header.Set("Authorization", c.AdminToken)
	}

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("error listando registros, status: %d", resp.StatusCode)
	}

	return io.ReadAll(resp.Body)
}

// CreateRecord crea un nuevo registro en una colección.
func (c *PocketBaseClient) CreateRecord(collection string, data interface{}) ([]byte, error) {
	endpoint := fmt.Sprintf("%s/api/collections/%s/records", c.BaseURL, collection)

	jsonData, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if c.AdminToken != "" {
		req.Header.Set("Authorization", c.AdminToken)
	}

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNoContent {
		bodyBytes, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("error creando registro, status: %d, response: %s", resp.StatusCode, string(bodyBytes))
	}

	return io.ReadAll(resp.Body)
}

// UpdateRecord actualiza un registro existente.
func (c *PocketBaseClient) UpdateRecord(collection, id string, data interface{}) ([]byte, error) {
	endpoint := fmt.Sprintf("%s/api/collections/%s/records/%s", c.BaseURL, collection, id)

	jsonData, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest("PATCH", endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if c.AdminToken != "" {
		req.Header.Set("Authorization", c.AdminToken)
	}

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		bodyBytes, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("error actualizando registro, status: %d, response: %s", resp.StatusCode, string(bodyBytes))
	}

	return io.ReadAll(resp.Body)
}

// DeleteRecord elimina un registro.
func (c *PocketBaseClient) DeleteRecord(collection, id string) error {
	endpoint := fmt.Sprintf("%s/api/collections/%s/records/%s", c.BaseURL, collection, id)

	req, err := http.NewRequest("DELETE", endpoint, nil)
	if err != nil {
		return err
	}
	if c.AdminToken != "" {
		req.Header.Set("Authorization", c.AdminToken)
	}

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNoContent {
		return fmt.Errorf("error eliminando registro, status: %d", resp.StatusCode)
	}

	return nil
}

File: backend/src/services/pocketbase.js
class PocketBaseClient {
  constructor(baseURL) {
    this.baseURL = baseURL;
    this.adminToken = null;
    this.httpClient = {
      timeout: 10000,
    };
  }

  async request(method, endpoint, body = null, headers = {}) {
    const url = new URL(endpoint, this.baseURL).toString();
    const opts = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      signal: AbortSignal.timeout(this.httpClient.timeout),
    };

    if (body) {
      opts.body = JSON.stringify(body);
    }

    const resp = await fetch(url, opts);
    return resp;
  }

  async authAdmin(email, password) {
    const resp = await this.request('POST', '/api/admins/auth-with-password', {
      identity: email,
      password,
    });

    if (!resp.ok) {
      throw new Error(`fallo la autenticación de admin, status: ${resp.status}`);
    }

    const result = await resp.json();
    this.adminToken = result.token;
  }

  async authUser(email, password) {
    const resp = await this.request('POST', '/api/collections/users/auth-with-password', {
      identity: email,
      password,
    });

    if (!resp.ok) {
      throw new Error(`credenciales inválidas, status: ${resp.status}`);
    }

    const result = await resp.json();
    return {
      token: result.token,
      record: {
        id: result.record.id,
        role: result.record.role,
      },
    };
  }

  async validateToken(token) {
    const resp = await this.request('POST', '/api/collections/users/auth-refresh', null, {
      Authorization: token,
    });

    if (!resp.ok) {
      throw new Error(`token inválido, PocketBase respondió status ${resp.status}`);
    }

    const result = await resp.json();
    return {
      id: result.record.id,
      role: result.record.role,
    };
  }

  async createUser(email, password, passwordConfirm, name, role) {
    const resp = await this.request('POST', '/api/collections/users/records', {
      email,
      password,
      passwordConfirm,
      name,
      role,
    });

    if (!resp.ok) {
      const body = await resp.text();
      throw new Error(`error creando usuario, status: ${resp.status}, response: ${body}`);
    }

    const record = await resp.json();
    return {
      id: record.id,
      role: record.role,
    };
  }

  async getRecord(collection, id) {
    const headers = {};
    if (this.adminToken) {
      headers.Authorization = this.adminToken;
    }

    const resp = await this.request('GET', `/api/collections/${collection}/records/${id}`, null, headers);

    if (!resp.ok) {
      throw new Error(`error obteniendo registro, status: ${resp.status}`);
    }

    return resp.json();
  }

  async listRecords(collection, page = 1, limit = 50, filter = '') {
    let url = `/api/collections/${collection}/records?page=${page}&perPage=${limit}`;
    if (filter) {
      url += `&filter=${encodeURIComponent(filter)}`;
    }

    const headers = {};
    if (this.adminToken) {
      headers.Authorization = this.adminToken;
    }

    const resp = await this.request('GET', url, null, headers);

    if (!resp.ok) {
      throw new Error(`error listando registros, status: ${resp.status}`);
    }

    return resp.json();
  }

  async createRecord(collection, data) {
    const headers = {};
    if (this.adminToken) {
      headers.Authorization = this.adminToken;
    }

    const resp = await this.request('POST', `/api/collections/${collection}/records`, data, headers);

    if (!resp.ok) {
      const body = await resp.text();
      throw new Error(`error creando registro, status: ${resp.status}, response: ${body}`);
    }

    return resp.json();
  }

  async updateRecord(collection, id, data) {
    const headers = {};
    if (this.adminToken) {
      headers.Authorization = this.adminToken;
    }

    const resp = await this.request('PATCH', `/api/collections/${collection}/records/${id}`, data, headers);

    if (!resp.ok) {
      const body = await resp.text();
      throw new Error(`error actualizando registro, status: ${resp.status}, response: ${body}`);
    }

    return resp.json();
  }

  async deleteRecord(collection, id) {
    const headers = {};
    if (this.adminToken) {
      headers.Authorization = this.adminToken;
    }

    const resp = await this.request('DELETE', `/api/collections/${collection}/records/${id}`, null, headers);

    if (!resp.ok) {
      throw new Error(`error eliminando registro, status: ${resp.status}`);
    }
  }
}

module.exports = PocketBaseClient;
